pysite/views/wiki/edit.py

```diff
@@ -26,16 +31,46 @@ def get(self, page):
         obj = self.db.get(self.table_name, page)
 
         if obj:
-            rst = obj["rst"]
-            title = obj["title"]
-            preview = obj["html"]
+            rst = obj.get("rst", "")
+            title = obj.get("title", "")
+            preview = obj.get("html", preview)
+
+            if obj.get("lock_expiry") and obj.get("lock_user") != self.user_data.get("user_id"):
+                lock_time = datetime.datetime.fromtimestamp(obj["lock_expiry"])
+                if datetime.datetime.utcnow() < lock_time:
+                    return self.render("wiki/page_in_use.html", page=page)
+
+        lock_expiry = datetime.datetime.utcnow() + datetime.timedelta(minutes=5)
+
+        # When debug mode is enabled, login is bypassed, meaning that user_data is not present.
+        if self.user_data:
+            self.db.insert(self.table_name, {
+                           "slug": page,
+                           "lock_expiry": lock_expiry.timestamp(),
+                           "lock_user": self.user_data.get("user_id")
+                           },
+                           conflict="update")
 
         return self.render("wiki/page_edit.html", page=page, rst=rst, title=title, preview=preview)
 
     @require_roles(*EDITOR_ROLES)
     @csrf
     def post(self, page):
-        rst = request.form["rst"]
+        before = self.db.get(self.table_name, page)
```

Comment:
This whole diffing-and-submitting work should be factored out to another function.

Reply:
Could probably do with better documentation/commenting as well